k2/python/csrc/array.cc

```diff
@@ -0,0 +1,141 @@
+// k2/python/csrc/array.cc
+
+// Copyright (c)  2020  Xiaomi Corporation (author: Haowen Qiu)
+
+// See ../../../LICENSE for clarification regarding multiple authors
+
+#include "k2/python/csrc/array.h"
+
+#include <memory>
+#include <utility>
+
+#include "k2/csrc/array.h"
+#include "k2/python/csrc/tensor.h"
+
+namespace k2 {
+/*
+   DLPackArray2 initializes Array2 with `cap_indexes` and `cap_data` which are
+   DLManagedTensors.
+
+   `cap_indexes` is usually a one dimension contiguous array, i.e.,
+   `cap_indexes.ndim == 1 && cap_indexes.strides[0] == 1`.
+
+   `cap_data` may have different shapes depending on `ValueType`:
+       1. if `ValueType` is a primitive type (e.g. `int32_t`), it will be
+          a one dimension contiguous array, i.e.,
+          `cap_data.ndim == 1 && cap_data.strides[0] == 1`.
+       2. if `ValueType` is a complex type (e.g. Arc), it will be a two
+          dimension array, i.e., it meets the following requirements:
+          a) cap_data.ndim == 2.
+          b) cap_data.shape[0] == num-elements it stores; note the
+             element's type is `ValueType`, which means we view each row of
+             `cap_data.data` as one element with type `ValueType`.
+          c) cap_data.shape[1] == num-primitive-values in `ValueType`,
+             which means we require that `ValueType` can be viewed as a tensor,
+             this is true for Arc as it only holds primitive values with same
+             type (i.e. `int32_t`), but may need type cast in other cases
+             (e.g. ValueType contains both `int32_t` and `float`).
+          d) cap_data.strides[0] == num-primitive-values in `ValueType`.
+          e) cap_data.strides[1] == 1.
+
+    Note if `data` in Array2 has stride > 1 (i.e. `data`'s type is
+    StridedPtr<ValueType>), the requirement of `cap_data` is nearly same with
+    case 2 above except cap_data.strides[0] will be greater than
+    num-primitive-values in `ValueType`.
+
+*/
+template <typename ValueType, bool IsPrimitive, typename I>
+class DLPackArray2;
+
+template <typename ValueType, typename I>
+class DLPackArray2<ValueType *, true, I> : public Array2<ValueType *, I> {
```

Comment:
The compiler should be able to figure out whether ValueType is primitive or not.
You do not need to pass true or false by yourself.

Reply:
How? there is a function `std::is_fundamental`, I can pass it as the value of `IsPrimitive`, but that also requires the template parameter.

Reply:
One possible method would be

```cpp
template<typename T> struct SupportedPrimitiveType : public std::false_type {};

template<> struct SupportedPrimitiveType<int32_t> : public std::true_type {};
template<> struct SupportedPrimitiveType<float> : public std::true_type {};

// add other supported types here
```

Then you can use

```cpp
template <typename ValueType, typename I>
class DLPackArray2;

// for primitive types
template<typename ValueType, typename I>
class DLPackArray2<typename std::enable_if<SupportedPrimitiveType<ValueType>::value, ValueType*>::type, I> {...};


// for non-primitive types

template<typename ValueType, typename I>
class DLPackArray2<typename std::enable_if<!SupportedPrimitiveType<ValueType>::value, ValueType*>::type, I> {...};

```

Reply:
mm..that's the same way with `std::is_fundamental`, you even do not need to write those `SupportedPrimitiveType` structs.

I prefer the simple one, as users will only expose a few types in `k2` (that's why I did not use `is_fundamental` here.)